src/legacy/reduce.rs

```diff
@@ -0,0 +1,292 @@
+use std::collections::VecDeque;
+use std::io::{self, Read, Result};
+
+use crate::legacy::lz77::lz77_output_backref;
+use bitstream_io::{BitRead, BitReader, Endianness, LittleEndian};
+
+/// Number of bits used to represent indices in a follower set of size n.
+fn follower_idx_bw(n: u8) -> u8 {
+    debug_assert!(n <= 32);
+    match n {
+        0 => 0,
+        1 => 1,
+        _ => 8 - (n - 1).leading_zeros() as u8,
+    }
+}
+
+#[derive(Default, Clone, Copy)]
+struct FollowerSet {
+    size: u8,
+    idx_bw: u8,
+    followers: [u8; 32],
+}
+
+/// Read the follower sets from is into fsets. Returns true on success.
+type FollowerSetArray = [FollowerSet; u8::MAX as usize + 1];
+
+fn read_follower_sets<T: std::io::Read, E: Endianness>(
+    is: &mut BitReader<T, E>,
+) -> io::Result<FollowerSetArray> {
+    let mut fsets = [FollowerSet::default(); u8::MAX as usize + 1];
+    for i in (0..=u8::MAX as usize).rev() {
+        let n = is.read::<u8>(6)?;
+        if n > 32 {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "invalid follower set",
+            ));
+        }
+        fsets[i].size = n;
+        fsets[i].idx_bw = follower_idx_bw(n);
+
+        for j in 0..fsets[i].size as usize {
+            fsets[i].followers[j] = is.read::<u8>(8)?;
+        }
+    }
+
+    Ok(fsets)
+}
+
+/// Read the next byte from is, decoded based on prev_byte and the follower sets.
+/// The byte is returned in *out_byte. The function returns true on success,
```

Comment:
No it doesn't, it returns Ok(u8) or Err(io::Error).

Reply:
fixed.